server/src/main/java/org/elasticsearch/action/search/SearchRequest.java

```diff
@@ -283,9 +283,15 @@ public ActionRequestValidationException validate() {
                 validationException = source.aggregations().validate(validationException);
             }
         }
-        if (reader() != null && reader().getId() != null) {
+        if (reader() != null) {
             if (indices.length > 0) {
-                validationException = addValidationError("[indices] cannot be used with reader contexts", validationException);
+                validationException = addValidationError("[index] cannot be used with reader contexts", validationException);
+            }
+            if (routing() != null) {
+                validationException = addValidationError("[routing] cannot be used with reader contexts", validationException);
+            }
+            if (preference() != null) {
+                validationException = addValidationError("[preference] cannot be used with reader contexts", validationException);
             }
```

Comment:
We should also check if scroll is not set ?

Reply:
addressed in https://github.com/elastic/elasticsearch/pull/55265/commits/b8e523de874d57a0c9572c513e5915fda5b596ad